src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -485,42 +477,69 @@ internal override Task FlushAsync(CancellationToken cancellationToken = default)
             return Task.CompletedTask;
         }
 
-        public override ValueTask DisposeAsync()
-        {
-            // TODO: perform a graceful shutdown and wait for completion?
+        public override ValueTask DisposeAsync(CancellationToken cancellationToken) =>
+            DisposeAsync(cancellationToken, async: true, immediate: _state.SendState == SendState.Aborted);
 
-            Dispose(true);
-            return default;
-        }
+        public override void Dispose() =>
+            Dispose(immediate: _state.SendState == SendState.Aborted);
 
-        public override void Dispose()
-        {
-            Dispose(true);
-            GC.SuppressFinalize(this);
-        }
+        ~MsQuicStream() =>
+            Dispose(immediate: true);
 
-        ~MsQuicStream()
+        private void Dispose(bool immediate)
         {
-            Dispose(false);
+            ValueTask t = DisposeAsync(cancellationToken: default, async: false, immediate);
+            Debug.Assert(t.IsCompleted);
+            t.GetAwaiter().GetResult();
         }
 
-        private void Dispose(bool disposing)
+        /// <param name="cancellationToken"></param>
+        /// <param name="async"></param>
+        /// <param name="immediate">When true, causes immediate disposal without waiting for peer ACKs.</param>
+        /// <returns></returns>
+        private async ValueTask DisposeAsync(CancellationToken cancellationToken, bool async, bool immediate)
         {
             if (_disposed)
             {
                 return;
             }
 
+            QUIC_STREAM_SHUTDOWN_FLAGS flags = immediate
+                ? (QUIC_STREAM_SHUTDOWN_FLAGS.GRACEFUL | QUIC_STREAM_SHUTDOWN_FLAGS.IMMEDIATE)
+                : QUIC_STREAM_SHUTDOWN_FLAGS.GRACEFUL;
+
+            StartShutdownOrAbort(flags, errorCode: 0);
+
+            if (async)
+            {
+                _state.RootedDisposeStream = this;
+                try
+                {
+                    await _state.ShutdownCompletionSource.Task.WaitAsync(cancellationToken).ConfigureAwait(false);
+                }
+                finally
+                {
+                    _state.RootedDisposeStream = null;
+                }
+            }
+            else
+            {
+                _state.ShutdownCompletionSource.Task.GetAwaiter().GetResult();
```

Comment:
If it was a bidi stream and peer did not send their shutdown, we'll wait here forever?

Reply:
Yes. The "recommended disposal behavior" in top comment is how we get around this right now.

The alternative is to make dispose be abortive, which we previously didn't like because other streams treat dispose as graceful.

Reply:
Looking at the "recommended disposal behavior" - should we then check if CloseAsync was called, which would mean we are on a recommended path -- and if not, perform an abortive dispose?

Reply:
Reading through the code again, I see that CloseAsync is just DisposeAsync, so the "recommended disposal behavior" is actually just putting a timeout on that wait via CancellationToken that you can't specify otherwise.

Reply:
The problem is that everyone else using `Stream` -- any stream, not just `QuicStream` or `NetworkStream` -- will assume that `Dispose()` will gracefully close the stream.

We'd be the odd one that made it abortive. So this is a workaround for now to keep consistent. I expect a long deliberation on this one once it gets into API review.